keras/layers/wrappers.py

```diff
@@ -133,3 +133,99 @@ def step(x, states):
             output_shape = self.get_output_shape_for(input_shape)
             y = K.reshape(y, (-1, input_length) + output_shape[2:])
         return y
+
+
+class Bidirectional(Wrapper):
+    ''' Bidirectional wrapper for RNNs
+
+    # Arguments:
+        layer: `Recurrent` instance.
+        merge_mode: Mode by which outputs of the forward and reverse RNNs will be combined. One of {sum, mul, concat, ave}
+
+    # Examples:
+    ```python
+    model = Sequential()
+    model.add(Bidirectional(LSTM(10, return_sequences=True), input_shape=(5, 10)))
+    model.add(Bidirectional(LSTM(10)))
+    model.add(Dense(5))
+    model.add(Activation('softmax'))
+    model.compile(loss='categorical_crossentropy', optimizer='rmsprop')
+    ```
+    '''
+    def __init__(self, layer, merge_mode='concat', weights=None, **kwargs):
+        self.forward = layer
+        self.reverse = layer.__class__.from_config(layer.get_config())
+        self.merge_mode = merge_mode
+        if weights:
+            nw = len(weights)
+            self.forward.initial_weights = weights[:nw//2]
+            self.reverse.initial_weights = weights[nw//2:]
+        self.stateful = layer.stateful
+        self.return_sequences = layer.return_sequences
+        self.supports_masking = True
+        super(Bidirectional, self).__init__(layer, **kwargs)
+
+    def get_weights(self):
+        return self.forward.get_weights() + self.reverse.get_weights()
+
+    def set_weights(self, weights):
+        nw = len(weights)
+        self.forward.set_weights(weights[:nw//2])
+        self.reverse.set_weights(weights[nw//2:])
+
+    def get_output_shape_for(self, input_shape):
+        if self.merge_mode in ['sum', 'ave', 'mul']:
+            return self.forward.get_output_shape_for(input_shape)
+        elif self.merge_mode == 'concat':
+            shape = list(self.forward.get_output_shape_for(input_shape))
+            shape[-1] *= 2
+            return tuple(shape)
+
+    def call(self, X, mask=None):
+
+        def reverse(x):
+            rev = K.permute_dimensions(x, (1, 0, 2))[::-1]                
```

Comment:
Currently that part of rnn is still buggy. For example in theano backend, with go_backwards=True and unroll=False, both the input and the mask is reversed, but if unroll=True, then only the input is reversed.